### PR TITLE
Test sync on testnet

### PR DESCRIPTION
**Pull request by teor2345**
These tests aren't enabled by default yet, because we need network access for sync tests in CI:
  - we have network on macOS, but not on Linux (and Windows was terminated due to the Linux failure, so we don't know if it has network access)
  - see #1141 for follow-up

TODO:
- [x]  rebase on main
- [x] fix a `TestDirExt` timeout bug
- [x] add tests for `TestDirExt`

### CODE REVIEW

**Comment by teor2345** [comment]
#1035 was closed because it was based on a merged branch.

(That's not actually what I wanted, but it seems to be a GitHub UI limitation.)

**Comment by teor2345** [comment]
I've tried enabling the sync tests by default, let's see if that works with the network access we currently have in CI.

**Review by yaahc — DISMISSED** [review]
Looks good, though I think I'd prefer to not add the `spawn_child_with_command` function and associated Trait. Unless you feel we will be spawning other cmds frequently I think it will be cleaner to just use `std::process::Command` directly + `CommandExt::spawn2` when we need to spawn something other than `zebrad` and leave this trait to be a zebrad specific helper.

**Comment by teor2345** [comment]
> Looks good, though I think I'd prefer to not add the `spawn_child_with_command` function and associated Trait. Unless you feel we will be spawning other cmds frequently I think it will be cleaner to just use `std::process::Command` directly + `CommandExt::spawn2` when we need to spawn something other than `zebrad` and leave this trait to be a zebrad specific helper.

I split `TestDirExt` and `ZebradTestDirExt` so that we can test `zebra_test::command` without running `zebrad`. Using `CommandExt::spawn2` doesn't help us test `ZebradTestDirExt`, because we aren't testing the code that's actually used by 
`ZebradTestDirExt`. (And trying to keep the code in sync manually would just lead to more bugs.)

See #1140 for a list of bugs I discovered in the original `TestDirExt` implementation. I fixed the timeout bug in this PR. This PR also documents the rest of the bugs in code comments, and #1140 helps us track fixing them (if they ever become a problem).

**Comment by teor2345** [comment]
@yaahc no wonder this PR didn't make sense, I forgot to add the command tests to the branch.

**Review by yaahc — DISMISSED** [review]
> @yaahc no wonder this PR didn't make sense, I forgot to add the command tests to the branch.

that makes a lot of sense, loool. Alright, looks good, wish we could avoid the vec and annoying slice manipulation but I can see why its needed now.

**Comment by teor2345** [comment]
It looks like the macOS tests have hung, maybe there's a bug or race condition in this PR which only shows up on macOS.

Unfortunately, I can't see the output until the tests finish.

**Comment by teor2345** [comment]
macOS has BSD `yes`, which treats `--version` as a string to repeat.
(Compared with GNU `yes`, which outputs its version.)

So that's why the initial "do we have this command?" check was hanging - `yes --version` never terminates on macOS.

Now I'm using `hexdump -v /dev/zero` as my "make lots of lines" command.

**Review by yaahc — APPROVED** [review]
LGTM

File: zebra-test/tests/command.rs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+            );
+            false
+        }
+        _ => true,
+    }
+}
+
+/// Test if a process that keeps on producing lines of output is killed after the timeout.
+#[test]
+fn kill_on_timeout_output_continuous_lines() -> Result<()> {
```
Comment (dconnolly) [inline zebra-test/tests/command.rs:44]: ❤️

File: zebra-test/tests/command.rs [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+    Ok(())
+}
+
+/// Test if the tests pass for a process that produces a single line of output,
+/// then exits before the timeout.
+//
+// TODO: create a similar test that pauses after output
+#[test]
+fn finish_before_timeout_output_single_line() -> Result<()> {
```
Comment (dconnolly) [inline zebra-test/tests/command.rs:74]: ❤️

File: zebra-test/tests/command.rs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    Ok(())
+}
+
+/// Test if a process that keeps on producing output, but doesn't produce any newlines,
+/// is killed after the timeout.
+///
+/// This test fails due to bugs in TestDirExt, see #1140 for details.
+#[test]
+#[ignore]
+fn kill_on_timeout_continuous_output_no_newlines() -> Result<()> {
```
Comment (dconnolly) [inline zebra-test/tests/command.rs:100]: ❤️

File: zebra-test/tests/command.rs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+    Ok(())
+}
+
+/// Test if tests pass for a process that produces a small amount of output,
+/// with no newlines, then exits before the timeout.
+//
+// TODO: create a similar test that pauses after output
+#[test]
+fn finish_before_timeout_short_output_no_newlines() -> Result<()> {
```
Comment (dconnolly) [inline zebra-test/tests/command.rs:125]: ❤️

File: zebra-test/tests/command.rs [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+    Ok(())
+}
+
+/// Test if the timeout works for a process that produces no output.
+///
+/// This test fails due to bugs in TestDirExt, see #1140 for details.
+#[test]
+#[ignore]
+fn kill_on_timeout_no_output() -> Result<()> {
```
Comment (dconnolly) [inline zebra-test/tests/command.rs:151]: ❤️

**Review by dconnolly — APPROVED** [review]
😍

**Pull request merged**